paf-mvp-core-js/src/model/audit-log.ts

```diff
@@ -0,0 +1,91 @@
+import {
+  AuditLog,
+  IdsAndPreferences,
+  Seed,
+  TransmissionRequest,
+  TransmissionResponse,
+  TransmissionResult,
+} from './generated-model';
+import { CurrentModelVersion } from './model';
+
+export const buildAuditLog = (
+  seed: Seed,
+  data: IdsAndPreferences,
+  response: TransmissionResponse,
+  contentId: string
+): AuditLog | undefined => {
+  if (response.version != CurrentModelVersion) {
+    return undefined;
+  }
+  const path = findTransactionPath(response, contentId);
+  if (path === undefined) {
+    return undefined;
+  }
+  return {
+    version: CurrentModelVersion,
+    data,
+    seed,
+    transaction_id: path.transactionId,
+    transmissions: path.results,
+  };
+};
+
+interface NodeTrack {
+  node: TransmissionResponse;
+  remainingChildren: TransmissionResponse[];
+}
+interface TransactionPath {
+  /** All the transmission results participating to the contentId, in order. */
+  results: TransmissionResult[];
+  transactionId: string;
+}
+
+const findTransactionPath = (response: TransmissionResponse, contentId: string): TransactionPath | undefined => {
+  const stack: NodeTrack[] = [];
+  let current: NodeTrack = { node: response, remainingChildren: [...response.children] };
+
+  while (current !== undefined || stack.length > 0) {
+    // Go to next leaf (depth search)
+    while (current != undefined) {
+      stack.push(current);
+      const content = current.node.contents.find((c) => c.content_id == contentId);
+      if (content !== undefined) {
+        return {
+          results: stack.map(fromResponseToResult),
+          transactionId: content.transaction_id,
+        };
+      }
+      if (current.remainingChildren.length > 0) {
+        const node = current.remainingChildren.pop();
+        current = { node, remainingChildren: [...node.children] };
```

Comment:
Is a copy of `node.children` really needed? can you not just do `remainingChildren: node.children`?

Reply:
We need a deep copy, unfortunately. We do a depth search in the tree, here. What may be unusual is that the tree isn't a binary one. So we cannot just do "left" and then "right", we need to keep track of the unvisited children and thus without modifying the existing tree. So we need to "attach" a list remaining children to each node that we are visiting.